src/elan/config.rs

```diff
@@ -231,14 +239,18 @@ impl Cfg {
             }
 
             // Then look for 'leanpkg.toml'
-            let toolchain_file = d.join("leanpkg.toml");
-            if let Ok(s) = utils::read_file("leanpkg.toml", &toolchain_file) {
-                // TODO(Sebastian): better error handling, override reason
-                let toml = s.parse::<toml::Value>().unwrap();
-                if let Some(s) = toml["package"].get("lean_version").and_then(toml::Value::as_str) {
-                    let toolchain_name = s.trim();
-                    let reason = OverrideReason::ToolchainFile(toolchain_file);
-                    return Ok(Some((toolchain_name.to_string(), reason)));
+            let leanpkg_file = d.join("leanpkg.toml");
+            if let Ok(content) = utils::read_file("leanpkg.toml", &leanpkg_file) {
+                let value = content.parse::<toml::Value>()
+                    .map_err(|error| ErrorKind::InvalidLeanpkgFile(leanpkg_file.clone(), error))?;
+                match value.get("package").and_then(|package| package.get("lean_version")) {
+                    None => {}
+                    Some(toml::Value::String(s)) => {
+                        return Ok(Some((s.to_string(), OverrideReason::LeanpkgFile(leanpkg_file))))
+                    }
+                    Some(a) => {
+                        return Err(ErrorKind::InvalidLeanVersion(leanpkg_file, a.type_str()).into())
+                    }
```

Comment:
Note that this is a change in behavior: previously, any non-string value
associated with `"package.lean_version"` was silently discarded.